src/udf/src/lib.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright 2023 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use arrow_array::RecordBatch;
+use arrow_flight::decode::FlightRecordBatchStream;
+use arrow_flight::encode::FlightDataEncoderBuilder;
+use arrow_flight::error::FlightError;
+use arrow_flight::flight_service_client::FlightServiceClient;
+use arrow_flight::{FlightData, FlightDescriptor};
+use arrow_schema::Schema;
+use futures_util::{stream, Stream, StreamExt, TryStreamExt};
+use tonic::transport::Channel;
+
+/// Client for external function service based on Arrow Flight.
+#[derive(Debug)]
+pub struct ArrowFlightUdfClient {
+    client: FlightServiceClient<Channel>,
+}
+
+impl ArrowFlightUdfClient {
+    /// Connect to a UDF service.
+    pub async fn connect(addr: &str) -> Result<Self> {
+        let client = FlightServiceClient::connect(addr.to_string()).await?;
+        Ok(Self { client })
+    }
+
+    /// Check if the function is available and return the function ID.
+    pub async fn check(&self, name: &str, args: &Schema, returns: &Schema) -> Result<FunctionId> {
+        let mut path = vec![name.to_string()];
+        for arg in &args.fields {
+            path.push(format!("{}", arg.data_type()));
+        }
+        let descriptor = FlightDescriptor::new_path(path.clone());
+
+        let response = self.client.clone().get_flight_info(descriptor).await?;
+
+        let info = response.into_inner();
+        let schema = Schema::try_from(info)
+            .map_err(|e| FlightError::DecodeError(format!("Error decoding schema: {e}")))?;
+        let expect_types: Vec<_> = returns.fields.iter().map(|f| f.data_type()).collect();
```

Comment:
There might be something special for the `Struct` type, which considers the names and the orders of the field for `Eq`. 🤔 I'm not sure what the correct behavior should be like.

Reply:
I'm not sure too. Have not tested it yet. I will add tests for more data types later.

Reply:
Looks like we don't even support `FromIntoArrow` for struct arrays. Let's leave the supports and fixes for `Struct` for the future.

Reply:
Actually we have. But let's leave it anyway. 😇